packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/FontAttributeProvider.kt

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+package com.facebook.react.views.text
+
+import android.content.res.AssetManager
+import android.graphics.Typeface
+
+interface FontAttributeProvider {
+  val fontFamily: String?
+
+  val fontStyle: Int
+
+  val fontWeight: Int
+}
+
+fun FontAttributeProvider.findEffectiveTypeface(
```

Comment:
I like splitting this out to be shared, but am wondering why this needs to be part of an inheritance hierarchy, if it's effectively just a pure function accepting some inputs?

Reply:
It doesn't need to be; I just implemented it this way.

Suppose you refactor this to be a non-method. In that case, you'll notice that both implementations pass their properties as the arguments, and these properties basically have the same name and (arguably) similar/same semantics.

I personally don't think about it as a "public" inheritance for the sake of this class' users, but more like a mix-in.

Existing programming languages are quite limiting in expressing such concepts, and often, an interface is the best thing you'll get.

In this case, the classes that implement this interface are "relatively" internal, so we've got the consequences under control, I think.

Reply:
This starts getting into territory of subjectivity, and this being structured as an interface is better than pure inheritance, but it means at the call-site of `findEffectiveTypeface`, we no longer know what state is actually an input to the algorithm, without looking up the interface et all.

In this case, where the state never changes over time,  we've kinda just taken `ReactTypefaceUtil`, curried a couple params to a new function. Imo the indirection/boilerplate makes the code less easy to quickly reason about, instead of more (esp for `ReactEditText`).

Reply:
> This starts getting into territory of subjectivity,

It sure does

> we no longer know what state is actually an input

It's nearly always the case with deduplication that the implementation details are taken away from you on the use site. You just assume that it does what it needs to do.

It can be argued both ways. When things are modeled as `Int`s or `String`s, as it's the case in this project, you could also argue that an explicitly modeled mixin interface brings benefits, as it gives confidence that the _right_ things are passed to the right places.

I was aware of the other way to model this, and I agree that it has some advantages. I decided to go this way.

Let me know if you consider this a blocker.

Reply:
I can add that the main reason I implemented this is that I add a third use later (if this changes anything).

Reply:
> I can add that the main reason I implemented this is that I add a third use later (if this changes anything).

What is the new usage of the interface?

Reply:
Sorry, I misinformed you. I ended up not adding a new implementation. **But** you can see how this interface is used in my work-in-progress implementation:

[(relevant usage)](https://github.com/cubuspl42/react-native/blob/41f212d590559c1ff9953cd37aaecbd6dfb5a8c7/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/span/ReactBackgroundSpan.kt#L28)

Reply:
@NickGerleman Did the real usage bring any new light here? Do you think we should "inline" the interface in [this place](https://github.com/cubuspl42/react-native/blob/41f212d590559c1ff9953cd37aaecbd6dfb5a8c7/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/span/ReactBackgroundSpan.kt#L47)?

Reply:
Also, could we clear out what you meant by...

> I like splitting this out to be shared [...]

...? If we nuke the `FontAttributeProvider` interface, this whole PR is close to being a no-op, so I think we'd need to close it in such a case.

Reply:
@NickGerleman Bump 🙂

Reply:
Yeah, I still don't love the interface here.